pypeline: add advanced invalidation

mrjackv opened the pull request.

mrjackv committed:
Add infrastructure for custom invalidation

Add infrastructure to pypeline to allow for pipes to provide custom
invalidation, by implementing an `invalidate` function.

mrjackv committed:
`Lift`: implement `invalidate`

mrjackv committed:
pipeline yaml: allow specifying pure analyses

Some analyses do not require any container to run, these can be now
specified as a standalone array.

mrjackv committed:
Introduce analyses that manipulate the model

Add 4 analyses to the pipebox that allow setting and verifying the model
and its diffs.

mrjackv committed:
Analysis: drop `name` argument in constructor

In the same vein as the `Pipe`'s counterpart, this feature is unused,
drop it for now.

mrjackv committed:
{Pipe,Analysis}Run: release Python GIL

Move out of the `runPipe`/`runAnalysis` functions the logic that
extracts the container from the vector type. This logic is now
delegated to `containerVectorToTuple` which converts the vector type to
a tuple of references. This allows python's `run{Pipe,Analysis}` to
safely call `nanobind::gil_scoped_release` which releases the GIL for
the entire duration of the `runPipe`/`runAnalysis` execution.

mrjackv committed:
LocalStorageProvider: use table for `model_paths`

When running `invalidate`, use a temporary table to store the
invalidated model paths that need to be used. This saves to template
them into the query and allows them to be recycled later when the
`dependency` table needs to be cleaned up.

mrjackv committed:
check-conventions: drop parenthesis detection

Drop a check in `check-conventions` that would mark commits as WIP if
they began with a parenthesis character (`([{`).

aleclearmind commented on libexec/revng/check-conventions:
```diff
@@ -636,7 +636,7 @@ class CheckCommitPass(Pass):
             problems.append("Commit subject line contains 'WIP'")
 
-        if subject[0] in "([{":
+        if (p_index := "([{".find(subject[0])) >= 0 and subject[-1] == ")]}"[p_index]:
```
1. Obscure code.
2. Why are we doing this? Why are we checking the last character?

mrjackv replied:
2. Otherwise check-conventions complains about the following subject `{Pipe,Analysis}Run: factor out container extract` because it begins with `{`

aleclearmind replied:
I still don't understand why you're checking the last character.

I'm not sure what you're trying to cover here.
This should prevent from using WIP in the commit title, I'd just prevent from using WIP entirely there. If you want we can also ban "fixup".

`\b(wip|fixup)\b`?

mrjackv replied:
> This should prevent from using WIP in the commit title

That's not the check in question, that one is the previous check which greps if there is `WIP` in the subject

> I still don't understand why you're checking the last character.

This check is intended for "temporary" commits, such as:
```
(do not push)
(End of commits from feature/XXXXX)
(SQUASH: XXXXX)
```

The check previously only checked the first character, where here I'm checking if the first character is `([{` **and** if there is a matching `)]}` at the end

aleclearmind replied:
I'm not sure why we have such check. The "temporary" commits you describe is not a well-known convention in revng. I'd drop it.

mrjackv replied:
Ok, I'll drop the check

mrjackv replied:
Dropped

aleclearmind commented on include/revng/PipeboxCommon/Helpers/Python/RunAnalysis.h:
```diff
@@ -18,14 +18,17 @@ inline llvm::Error runAnalysis(T &Handle,
-                                               Configuration,
-                                               Containers);
+  {
+    nanobind::gil_scoped_release X;
```
This commit is not about "factoring out" something, is about introducing `gil_scoped_release`.

mrjackv replied:
Reworked the commit message to reflect this

aleclearmind replied:
The message now says "allow". Please be super-explicit that we're introducing the lock.

Something like `{Pipe,Analysis}Run: release Python GIL`.

Touching locks/multi-threading is something very delicate, put the right weight on it.

mrjackv replied:
Reworked the commit message to be extra clear on what's happening

aleclearmind commented on include/revng/PipeboxCommon/Helpers/Helpers.h:
```diff
@@ -4,13 +4,22 @@
-struct ExtractContainerFromList {
-  static C &get(ListType &Containers);
-};
+/// Helper fuVnction that simplifies unpacking pipe's container list into a
```
Typo

mrjackv replied:
Fixed

aleclearmind commented on include/revng/PipeboxCommon/Helpers/Helpers.h:
```diff
@@ -4,13 +4,22 @@
+/// `runAnalysis`.
+template<typename T, typename CallableT>
+  requires(IsPipe<T> or IsAnalysis<T>)
+auto makeContainerTuple(CallableT &&Callable) {
```
1. `makeContainerTuple` is not a great name, it should be more explicit about the fact that it builds a tuple *given a pipe* (or an analysis).
2. The fact that you use the same name for this and the other two functions makes the code obscure.
3. Try to rewrite this with more decoupled:

```c++
// Implement for vector<Container *> and for nanobind::list
template<typename ContainerType>
struct ExtractFromContainerTrait {
  template<size_t Index, typename ExpectedType>
  ExpectedType &get(ContainerType &C);
};

template<IsTupleOfReferences Tuple,
         typename ContainerType>
Tuple vectorToTuple(ContainerVector &Containers) {
  // I know you can't do this here, but with an helper function
  // where you expand the Is, it should be doable.
  return Tuple(ExtractFromContainerTrait::get<I, std::tuple_element<I, Tuple>>(Containers)...);
}
```

Advantages: a single `vectorToTuple` function, no lambdas, decouple from the analysis/pipes.
Then you'll need to make a function that given a pipe/analysis calls this with the appropriate tuple.

mrjackv replied:
Used `containerVectorToTuple` for the function name, recycled the `ExtractContainerFromList` struct that was previously used by `runPipe` and `runAnalysis`.

aleclearmind commented on python/revng/pypeline/schema.yml:
```diff
@@ -17,6 +17,18 @@ $defs:
+            analysis:
+              type: string
+              description: The name of the analysis to run.
+        minItems: 1
```
I'd drop `minItems: 1`.

mrjackv replied:
Dropping it is marginal, it allows writing `analyses: []` and it's not consistent with the rest of the schema

aleclearmind replied:
I'd drop all `minItems: 1`. Schemas are supposed to validate semantics and having zero analysis is perfectly fine. Schemas are not for enforcing linting rules.

mrjackv replied:
Dropped `minItems`

aleclearmind commented on python/revng/pypeline/schema.yml:
```diff
@@ -17,6 +17,18 @@ $defs:
         uniqueItems: true
         items:
           $ref: "#/$defs/container_decl"
+      analyses:
```
Maybe we should call these "pure analyses" or "inputless analyses".

If we just say analyses one might think he can find all of them here.

mrjackv replied:
Given what you said below:

> This is (will be) a private schema, no need to think to backward compatibility.

I think it makes more sense to leave it as `analyses`. There's already schema constraints/assertions in place that forbid this from having containers and the pipe/savepoint ones from not having any

aleclearmind commented on python/revng/pypeline/schema.yml:
```diff
@@ -17,6 +17,18 @@ $defs:
+          properties:
+            analysis:
+              type: string
+              description: The name of the analysis to run.
```
Why not just a list of strings?

mrjackv replied:
1. Just in case in the future we need to add extra keys, e.g. `Docs` we don't have to convert the list of strings into a list of objects
2. It allows re-cycling the parsing logic for the non-container analyses

aleclearmind replied:
This is (will be) a private schema, no need to think to backward compatibility.

To me it's needlessly cumbersome, if we add stuff we'll fix this, let's not do thing just in case.

mrjackv replied:
Had to re-add the `name` key, closing since this comment is no longer valid.

aleclearmind commented on python/revng/internal/pipeline.yml:
```diff
@@ -28,6 +28,12 @@ containers:
+  - analysis: ApplyDiff
+  - analysis: VerifyDiff
+  - analysis: SetModel
+  - analysis: VerifyModel
```
kebab-case?

mrjackv replied:
Pipes are so far `PascalCase`, I'd stick with that since it reflects the class name of the analysis as well.

aleclearmind replied:
These become available on the CLI, and there we must use kebab-case.

mrjackv replied:
Re-added the `name` key, analyses are now `kebab-case`.

aleclearmind replied:
The name of the analysis itself should be kebab-case.
Let's keep the number of names to the minimum.
`VerifyModel` is the name of the `class`. The analysis is called `verify-model` and this should be the `Name` field as well.

mrjackv replied:
That means also having the python-generated nanbind class name be `kebab-case` which is not valid python
Addendum: also pipes currently have `PascalCase` names, and those will show up when doing `revng2 pipeline run-pipe`, should those also be `kebab-case`?

aleclearmind replied:
That's a distinct problem that has to be resolved when we register a pipe/analysis, possibly with an additional string or by making `RegisterPipe` a macro.

The name of the analysis out of Python and C++ (i.e., the CLI) has to be kebab-case and it's not a property of the pipeline, it's just how things must be.

mrjackv replied:
Analysis names now reflect the `Name` attribute in C++, which is `kebab-case`

aleclearmind commented on include/revng/PipeboxCommon/ModelAnalyses.h:
```diff
@@ -0,0 +1,73 @@
+
+#include "revng/PipeboxCommon/Model.h"
+
+struct ApplyDiff {
```
Use a `class`. We use `struct` for stuff we want to have all fields public, typically simple data structures.

mrjackv replied:
Done

aleclearmind commented on include/revng/PipeboxCommon/ModelManipulationAnalyses.h:
1. Could we implement these analyses in Python? Not necessarily today.
2. `s/ModelAnalyses/BasicAnalyses/` or `ModelManipulationAnalyses`.

mrjackv replied:
1. Hypotetically yes, but it would be very expensive because the model would need to be serialized, re-parsed in python, the modifications made and then re-parsed back in C++
2. Ok

mrjackv replied:
2. Renamed to `ModelManipulationAnalyses`

aleclearmind commented on include/revng/PipeboxCommon/ModelAnalyses.h:
```diff
@@ -0,0 +1,73 @@
+  }
+};
+
+struct VerifyModel {
```
What do we use this one for?

mrjackv replied:
The `Verify*` analyses are there in case someone that doesn't have access to revng directly (e.g. `revng-ui`, scripting) needs to verify that a model/diff is valid

aleclearmind replied:
Can't they just try to apply it and report back the error in case of failure?

mrjackv replied:
The semantic is different:
* `ApplyDiff`: I want to apply the diff and check the error if it doesn't apply (will modify the model if the diff is well-formed)
* `VerifyDiff`: would this diff hypotetically apply to this model (does not actually modify the model if the diff is well-formed)
* `SetModel`: I have a model file, please make it the model of the project
* `VerifyModel`: verify the provided model (basically `model::verify` but can be called from the daemon)

The semantic is of `ApplyDiff` and `SetModel` is to modify the project if everything is well-formed, this is not the same as `Verify{Diff,Model}` which do guarantee that no matter what the current model is not changed.

aleclearmind replied:
Do we actually use these in the UI? If so, why?

mrjackv replied:
We don't, but not having them also means that potential users of scripting do not have access to them.

aleclearmind replied:
We should eventually devise a better solution to this. This is an hack.

Add a comment about this.

mrjackv replied:
Added

aleclearmind commented on include/revng/PipeboxCommon/ModelAnalyses.h:
```diff
@@ -0,0 +1,73 @@
+  }
+};
+
+struct VerifyDiff {
```
What do we use this one for?

mrjackv replied:
See answer on the previous comment

aleclearmind commented on include/revng/PipeboxCommon/Common.h:
```diff
@@ -78,4 +74,22 @@ class Buffer {
+  std::vector<std::tuple<ObjectID, Buffer>>>;
+
+/// Type representing the output of a `Pipe::run` method
+using PipeOutput = std::tuple<ObjectDependencies, AdvancedInvalidation>;
```
Switch to `struct`s, no point in using pairs like this.

mrjackv replied:
Done

aleclearmind commented on include/revng/PipeboxCommon/Common.h:
```diff
@@ -78,4 +74,22 @@ class Buffer {
+/// Type used to return advanced invalidation data from a pipe run. Will be fed
+/// back to the pipe's `invalidate` function to determine if there are
+/// additional objects to invalidate.
+using AdvancedInvalidation = std::vector<
```
Let's drop "advanced" entirely in favor of "custom". For instance, `CustomInvalidationData` (I know I don't like data, but here the content is opaque, so I don't dislike it too much).

Explain why `vector<vector<...>>`.

mrjackv replied:
Done

aleclearmind commented on include/revng/PipeboxCommon/Helpers/PipeRunPipes/Base.h:
```diff
@@ -58,6 +58,26 @@ struct CheckPreconditionMixin<T> {
+struct InvalidateMixin {};
+
+template<HasPipeRunInvalidate T>
+struct InvalidateMixin<T> {
```
This only seems to have a single usage, you can inline it and use `requires` instead.

```c++
template<typename T>
class Asd {
    int lol() requires false { return 42; }
};
```

mrjackv replied:
Done

aleclearmind commented on python/revng/pypeline/model.py:
```diff
@@ -15,6 +15,20 @@
 
+class ModelDiff(ABC):
+    """
+    Abstract class representing a diff between two model classes
```
two model instances

mrjackv replied:
Fixed

aleclearmind commented on python/revng/internal/pipebox.py:
```diff
@@ -57,9 +57,9 @@ def run(
         outgoing: list[ObjectSet],
         configuration: Configuration,
-    ) -> PipeObjectDependencies:
+    ) -> PipeRunOutput:
```
`PipeRunOutput` is too empty as a term.

What about just `PipeDependencies`?

mrjackv replied:
Renamed

aleclearmind commented on python/revng/pypeline/storage/local_provider.py:
```diff
@@ -504,3 +557,40 @@ def _compare_file(path: Path, request: FileRequest) -> bool:
+                    )
+            self._write_metadata(cursor)
+
+    def get_advanced_dependencies(
```
This does not return dependencies but opaque data that enable you to compute dependencies.

`get_custom_invalidation_data`

mrjackv replied:
Renamed to `add_custom_invalidation_data` and `get_custom_invalidation_data`

aleclearmind commented on python/revng/pypeline/pipeline.py:
```diff
@@ -440,7 +441,39 @@ def run_analysis(
+
+        diff = model.diff(ReadOnlyModel(new_model))
+
+        adv_invalidated_objects: list[ObjectsToInvalidate] = []
```
1. No abbreviations
2. Outline the handling of custom invalidation in a function, makes the function more readable

mrjackv replied:
1. Done
2. Done

aleclearmind commented on python/revng/pypeline/pipeline.py:
```diff
@@ -440,7 +441,39 @@ def run_analysis(
+            # Optimization: since few pipes actually implement advanced
+            # invalidation and it requires querying storage, check that the
+            # method has actually been overridden.
+            if node.task.__class__.invalidate is Pipe.invalidate:
```
Hide this in `pipe.has_custom_invalidation`

mrjackv replied:
Done

aleclearmind commented on python/revng/pypeline/pipeline_node.py:
```diff
@@ -236,7 +237,21 @@ def run(
                 result.extend((self.bindings[index].name, obj, path) for obj, path in index_deps)
-            return result
+
+            if not all(len(x) == 0 for x in pipe_output.advanced_invalidation):
```
Comment

mrjackv replied:
Added comments throughout

aleclearmind commented on python/revng/pypeline/task/pipe.py:
```diff
@@ -2,15 +2,57 @@
+@dataclass
+class PipeOutput:
+    dependencies: ObjectDependencies
+    advanced_invalidation: PipeAdvancedInvalidation = field(default_factory=list)
```
```
PipeOutput -> ScheduledTaskDependencies
PipeRunOutput -> PipeDependencies
```

mrjackv replied:
Renamed

aleclearmind commented on include/revng/Lift/Lift.h:
```diff
@@ -176,11 +176,17 @@ class LiftPass : public llvm::ModulePass {
 };
 
-namespace detail {
+namespace lift::internal {
```
Our namespaces, except for `detail`, always start with `revng::`.

mrjackv replied:
Added `revng::`

aleclearmind commented on lib/Lift/Lift.cpp:
```diff
@@ -102,13 +102,28 @@ bool LiftPass::runOnModule(llvm::Module &M) {
 
+using JumpTargetMap = std::map<MetaAddress, bool>;
+
+template<>
+struct llvm::yaml::CustomMappingTraits<JumpTargetMap> {
+  static void inputOne(IO &IO, StringRef Key, JumpTargetMap &Data) {
+    MetaAddress Address = MetaAddress::fromString(Key);
+    IO.mapRequired(Key.str().c_str(), Data[Address]);
+  }
+
+  static void output(IO &IO, JumpTargetMap &Data) {
+    for (auto &[Key, Value] : Data)
+      IO.mapRequired(Key.toString().c_str(), Value);
+  }
+};
```
We have `llvm::yaml::ScalarTraits` for `MetaAddress`, I don't think you need to handle this by hand with `CustomMappingTraits`.

mrjackv replied:
The traits are not for `MetaAddress`, they are for the whole `std::map<MetaAddress, bool>` class. I can look to dispatch the `MetaAddress` logic to the `ScalarTraits` but it looks a lot of extra adapting for marginal gains

aleclearmind replied:
OK

aleclearmind commented:
@mrjackv please add a test for the advanced/custom invalidation feature in Python.

aleclearmind commented on lib/Lift/Lift.cpp:
```diff
@@ -100,3 +101,106 @@ bool LiftPass::runOnModule(llvm::Module &M) {
   return false;
 }
+
+using JumpTargetMap = std::map<MetaAddress, bool>;
```
What's the semantics of the `bool`? Docs.

mrjackv replied:
Added doxygen

mrjackv committed:
get_registry: use `name` for selected classes

Use the `name` class attribute to build the registry when discovering
subclasses of `Pipe`, `Analysis` and `Container`.

mrjackv committed:
pypeline: convert pipe names to `kebab-case`

mrjackv committed:
Add test for custom invalidation

Add a test to the pypeline tests that checks that the pipeline actually
invalidates the objects communicated by the custom invalidation of a
pipe.

mrjackv commented:
> @mrjackv please add a test for the advanced/custom invalidation feature in Python.

Done

aleclearmind commented on tests/pypeline/pipebox.py:
```diff
@@ -532,6 +533,65 @@ def run(
+        containers: list[Container],
+        incoming: list[ObjectSet],
+        outgoing: list[ObjectSet],
+        configuration: Configuration,
+    ) -> PipeDependencies:
+        object_ = MyObjectID(MyKind.ROOT)
+        container = cast(RootDictContainer, containers[0])
+        container.add_object(object_)
+        return PipeDependencies([[]], [[(object_, b"foo")]])
+
+    def invalidate(
+        self, invalidation_data: PipeCustomInvalidation, diff: ModelDiff
+    ) -> list[ObjectSet]:
+        assert len(invalidation_data) == 1
+        assert len(invalidation_data[0]) == 1
+        object_, data = invalidation_data[0][0]
+        assert object_ == MyObjectID(MyKind.ROOT)
+        assert data == b"foo"
+        return [ObjectSet(MyKind.ROOT, {object_})]
```
Comment

mrjackv replied:
Done

aleclearmind commented on tests/pypeline/pipebox.py:
```diff
@@ -532,6 +533,65 @@ def run(
+        incoming: list[ObjectSet],
+        configuration: str,
+    ):
+        # This analysis does nothing
```
Comment why

mrjackv replied:
Done

aleclearmind commented on tests/pypeline/test_all.py:
```diff
@@ -707,3 +708,40 @@ def add_object(save_start: int, save_end: int, object_id, path: str):
     }
+
+
+def test_custom_invalidation(model, storage_provider: StorageProvider):
```
Comment this function

mrjackv replied:
Done

mrjackv committed:
Artifact: fix typo in `write_to_disk`

Fix a typo in `write_to_disk` where the file open mode was set to `rb`
when it should have been `wb`.
